File: adapters/graphql-ts/test-adapter.ts
/**
 * Integration tests for the ANIP GraphQL adapter (TypeScript).
 *
 * Usage:
 *   npx tsx test-adapter.ts http://localhost:9100
 */

import { loadConfig } from "./src/config.js";
import type { AdapterConfig } from "./src/config.js";
import { discoverService } from "./src/discovery.js";
import { ANIPInvoker, CredentialError } from "./src/invocation.js";
import { generateSchema, toCamelCase, toSnakeCase } from "./src/translation.js";
import { buildApp } from "./src/index.js";
import { buildSchema } from "graphql";

// ---------------------------------------------------------------------------
// Test helpers
// ---------------------------------------------------------------------------

let passCount = 0;
let failCount = 0;

function ok(label: string): void {
  passCount++;
  console.log(`  PASS  ${label}`);
}

function fail(label: string, detail = ""): void {
  failCount++;
  let msg = `  FAIL  ${label}`;
  if (detail) msg += `  \u2014 ${detail}`;
  console.log(msg);
}

function assert(condition: boolean, label: string, detail = ""): void {
  if (condition) ok(label);
  else fail(label, detail);
}

// ---------------------------------------------------------------------------
// 1. Config defaults
// ---------------------------------------------------------------------------

function testConfigDefaults(): void {
  console.log("\n--- Config defaults ---");

  const oldUrl = process.env.ANIP_SERVICE_URL;
  const oldPort = process.env.ANIP_ADAPTER_PORT;
  const oldConfig = process.env.ANIP_ADAPTER_CONFIG;

  delete process.env.ANIP_SERVICE_URL;
  delete process.env.ANIP_ADAPTER_PORT;
  delete process.env.ANIP_ADAPTER_CONFIG;

  const cfg = loadConfig("/nonexistent/path.yaml");

  if (oldUrl) process.env.ANIP_SERVICE_URL = oldUrl;
  if (oldPort) process.env.ANIP_ADAPTER_PORT = oldPort;
  if (oldConfig) process.env.ANIP_ADAPTER_CONFIG = oldConfig;

  assert(
    cfg.anipServiceUrl === "http://localhost:8000",
    "default service URL",
    `got ${cfg.anipServiceUrl}`
  );
  assert(cfg.port === 3002, "default port", `got ${cfg.port}`);
  assert(
    cfg.graphqlPath === "/graphql",
    "default graphql path",
    `got ${cfg.graphqlPath}`
  );
}

// ---------------------------------------------------------------------------
// 2. Discovery
// ---------------------------------------------------------------------------

async function testDiscovery(anipUrl: string): Promise<void> {
  console.log("\n--- Discovery ---");

  const service = await discoverService(anipUrl);
  assert(service.baseUrl !== "", "base_url populated");
  assert(service.protocol !== "", "protocol populated");
  assert(service.capabilities.size > 0, "capabilities discovered");
  assert(service.capabilities.has("search_flights"), "search_flights found");
  assert(service.capabilities.has("book_flight"), "book_flight found");

  const sf = service.capabilities.get("search_flights")!;
  assert(sf.sideEffect === "read", "search_flights is read");
  assert(sf.financial === false, "search_flights not financial");

  const bf = service.capabilities.get("book_flight")!;
  assert(
    bf.sideEffect === "write" || bf.sideEffect === "irreversible",
    "book_flight has side effect"
  );
  assert(bf.financial === true, "book_flight is financial");
  assert(bf.requires.length > 0, "book_flight has prerequisites");
}

// ---------------------------------------------------------------------------
// 3. Schema generation
// ---------------------------------------------------------------------------

async function testSchemaGeneration(anipUrl: string): Promise<void> {
  console.log("\n--- Schema generation ---");

  const service = await discoverService(anipUrl);
  const sdl = generateSchema(service);

  // Directives present
  assert(sdl.includes("directive @anipSideEffect"), "SDL has @anipSideEffect directive");
  assert(sdl.includes("directive @anipCost"), "SDL has @anipCost directive");
  assert(sdl.includes("directive @anipRequires"), "SDL has @anipRequires directive");
  assert(sdl.includes("directive @anipScope"), "SDL has @anipScope directive");

  // Shared types
  assert(sdl.includes("scalar JSON"), "SDL has JSON scalar");
  assert(sdl.includes("type CostActual"), "SDL has CostActual type");
  assert(sdl.includes("type ANIPFailure"), "SDL has ANIPFailure type");
  assert(sdl.includes("type Resolution"), "SDL has Resolution type");

  // Query and Mutation
  assert(sdl.includes("type Query"), "SDL has Query type");
  assert(sdl.includes("type Mutation"), "SDL has Mutation type");

  // camelCase names
  assert(sdl.includes("searchFlights"), "SDL uses camelCase searchFlights");
  assert(sdl.includes("bookFlight"), "SDL uses camelCase bookFlight");

  // Result types (PascalCase)
  assert(sdl.includes("SearchFlightsResult"), "SDL has SearchFlightsResult type");
  assert(sdl.includes("BookFlightResult"), "SDL has BookFlightResult type");

  // searchFlights is in Query (read), bookFlight is in Mutation
  const queryStart = sdl.indexOf("type Query");
  const queryEnd = sdl.indexOf("}", queryStart);
  const queryBlock = sdl.substring(queryStart, queryEnd + 1);
  assert(queryBlock.includes("searchFlights"), "searchFlights is in Query type");

  const mutationStart = sdl.indexOf("type Mutation");
  const mutationEnd = sdl.indexOf("}", mutationStart);
  const mutationBlock = sdl.substring(mutationStart, mutationEnd + 1);
  assert(mutationBlock.includes("bookFlight"), "bookFlight is in Mutation type");

  // Directives applied to fields
  const afterQuery = sdl.substring(queryStart);
  assert(afterQuery.includes("@anipSideEffect"), "fields have @anipSideEffect");

  // Verify schema can be built with buildSchema()
  try {
    buildSchema(sdl);
    ok("SDL can be parsed by buildSchema()");
  } catch (e) {
    fail("SDL can be parsed by buildSchema()", String(e));
  }
}

// ---------------------------------------------------------------------------
// 4. Invocation
// ---------------------------------------------------------------------------

async function testInvocation(anipUrl: string): Promise<void> {
  console.log("\n--- Invocation ---");

  const service = await discoverService(anipUrl);
  const invoker = new ANIPInvoker(service);

  // API-key path: search flights
  const result = await invoker.invoke(
    "search_flights",
    {
      origin: "SEA",
      destination: "SFO",
      date: "2026-03-10",
      passengers: 1,
    },
    { apiKey: "demo-human-key" },
  );
  assert(typeof result === "object", "search_flights returns dict");
  assert(result.success === true, "search_flights succeeds");
  assert("result" in result, "search_flights has result");

  // API-key path: book flight
  const bookResult = await invoker.invoke(
    "book_flight",
    {
      flight_number: "AA100",
      date: "2026-03-10",
      passengers: 1,
    },
    { apiKey: "demo-human-key" },
  );
  assert(typeof bookResult === "object", "book_flight returns dict");
  assert("success" in bookResult, "book_flight has success key");

  // Token path: pre-issue a token, then invoke with it
  const tokenResp = await fetch(service.endpoints.tokens, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Authorization": "Bearer demo-human-key",
    },
    body: JSON.stringify({
      subject: "adapter:anip-graphql-adapter-ts",
      scope: ["travel.search"],
      capability: "search_flights",
    }),
  });
  assert(tokenResp.ok, "token request succeeded");
  const tokenData = (await tokenResp.json()) as Record<string, unknown>;
  assert(tokenData.issued === true, "token issued for token-path test");
  const jwt = tokenData.token as string;

  const tokenResult = await invoker.invoke(
    "search_flights",
    {
      origin: "SEA",
      destination: "SFO",
      date: "2026-03-10",
      passengers: 1,
    },
    { token: jwt },
  );
  assert(typeof tokenResult === "object", "token-path search_flights returns dict");
  assert(tokenResult.success === true, "token-path search_flights succeeds");

  // No credentials raises CredentialError
  try {
    await invoker.invoke(
      "search_flights",
      { origin: "SEA", destination: "SFO", date: "2026-03-10", passengers: 1 },
      {},
    );
    fail("no-creds raises CredentialError", "no exception raised");
  } catch (e) {
    if (e instanceof CredentialError) {
      ok("no-creds raises CredentialError");
    } else {
      fail(
        "no-creds raises CredentialError",
        `wrong exception: ${e instanceof Error ? e.constructor.name : String(e)}`
      );
    }
  }
}

// ---------------------------------------------------------------------------
// 5. Full server (Hono app.request)
// ---------------------------------------------------------------------------

async function testServer(anipUrl: string): Promise<void> {
  console.log("\n--- Server (Hono + graphql-js) ---");

  const config: AdapterConfig = {
    anipServiceUrl: anipUrl,
    port: 0,
    graphqlPath: "/graphql",
  };

  const app = await buildApp(config);

  // GET /schema.graphql returns SDL
  let resp = await app.request("/schema.graphql");
  assert(resp.status === 200, "GET /schema.graphql returns 200");
  const sdl = await resp.text();
  assert(sdl.includes("type Query"), "schema.graphql contains Query type");
  assert(
    sdl.includes("directive @anipSideEffect"),
    "schema.graphql has directives"
  );

  // Missing credentials test: POST without headers should return failure
  resp = await app.request("/graphql", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      query: `query {
        searchFlights(
          origin: "SEA"
          destination: "SFO"
          date: "2026-03-10"
        ) {
          success
          failure { type detail }
        }
      }`,
    }),
  });
  assert(resp.status === 200, "POST /graphql without creds returns 200 (GraphQL envelope)");
  let data = (await resp.json()) as Record<string, unknown>;
  const noCreds = ((data.data as Record<string, unknown>) ?? {})
    .searchFlights as Record<string, unknown> | undefined;
  assert(noCreds?.success === false, "missing-creds searchFlights success is false");
  const noCredsFailure = noCreds?.failure as Record<string, unknown> | undefined;
  assert(
    noCredsFailure?.type === "missing_credentials",
    "missing-creds failure type is missing_credentials",
    `got ${noCredsFailure?.type}`
  );

  // API-key path: POST /graphql with searchFlights query
  resp = await app.request("/graphql", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-ANIP-API-Key": "demo-human-key",
    },
    body: JSON.stringify({
      query: `query {
        searchFlights(
          origin: "SEA"
          destination: "SFO"
          date: "2026-03-10"
        ) {
          success
          result
        }
      }`,
    }),
  });
  assert(resp.status === 200, "POST /graphql searchFlights returns 200");
  data = (await resp.json()) as Record<string, unknown>;
  const errors = data.errors as unknown[] | undefined;
  assert(
    !errors || errors.length === 0,
    `searchFlights has no errors`,
    JSON.stringify(errors)
  );
  const searchData = ((data.data as Record<string, unknown>) ?? {})
    .searchFlights as Record<string, unknown> | undefined;
  assert(searchData?.success === true, "searchFlights success is true");
  assert(searchData?.result != null, "searchFlights has result");

  // Token path: pre-issue a signed token via real fetch to ANIP server
  const tokenResp = await fetch(`${anipUrl}/anip/tokens`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Authorization": "Bearer demo-human-key",
    },
    body: JSON.stringify({
      subject: "adapter:anip-graphql-adapter-ts",
      scope: ["travel.search"],
      capability: "search_flights",
    }),
  });
  const tokenData = await tokenResp.json() as Record<string, unknown>;
  assert(tokenData.issued === true, "token issued for token-path test");
  const signedToken = tokenData.token as string;

  resp = await app.request("/graphql", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-ANIP-Token": signedToken,
    },
    body: JSON.stringify({
      query: `query {
        searchFlights(
          origin: "SEA"
          destination: "SFO"
          date: "2026-03-10"
        ) {
          success
          result
        }
      }`,
    }),
  });
  assert(resp.status === 200, "POST /graphql with token returns 200");
  data = (await resp.json()) as Record<string, unknown>;
  const tokenErrors = data.errors as unknown[] | undefined;
  assert(
    !tokenErrors || tokenErrors.length === 0,
    `token-path searchFlights has no errors`,
    JSON.stringify(tokenErrors)
  );
  const tokenSearchData = ((data.data as Record<string, unknown>) ?? {})
    .searchFlights as Record<string, unknown> | undefined;
  assert(tokenSearchData?.success === true, "token-path searchFlights success is true");

  // POST /graphql with bookFlight mutation
  resp = await app.request("/graphql", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-ANIP-API-Key": "demo-human-key",
    },
    body: JSON.stringify({
      query: `mutation {
        bookFlight(
          flightNumber: "AA100"
          date: "2026-03-10"
          passengers: 1
        ) {
          success
          result
          costActual {
            financial {
              amount
              currency
            }
            varianceFromEstimate
          }
        }
      }`,
    }),
  });
  assert(resp.status === 200, "POST /graphql bookFlight returns 200");
  data = (await resp.json()) as Record<string, unknown>;
  const bookErrors = data.errors as unknown[] | undefined;
  assert(
    !bookErrors || bookErrors.length === 0,
    `bookFlight has no errors`,
    JSON.stringify(bookErrors)
  );
  const bookData = ((data.data as Record<string, unknown>) ?? {})
    .bookFlight as Record<string, unknown> | undefined;
  assert(bookData != null && "success" in bookData, "bookFlight has success key");

  // GET /graphql returns HTML playground
  resp = await app.request("/graphql");
  assert(resp.status === 200, "GET /graphql returns 200 (playground)");
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main() {
  if (process.argv.length < 3) {
    console.log("Usage: npx tsx test-adapter.ts <anip-service-url>");
    console.log("Example: npx tsx test-adapter.ts http://localhost:9100");
    process.exit(1);
  }

  const anipUrl = process.argv[2];
  console.log(`Testing ANIP GraphQL adapter (TS) against ${anipUrl}`);

  // 1. Config (sync)
  testConfigDefaults();

  // 2-5. Async tests
  await testDiscovery(anipUrl);
  await testSchemaGeneration(anipUrl);
  await testInvocation(anipUrl);
  await testServer(anipUrl);

  // Summary
  console.log(`\n${"=".repeat(50)}`);
  console.log(`Results: ${passCount} passed, ${failCount} failed`);
  if (failCount > 0) {
    process.exit(1);
  }
  console.log("All tests passed!");
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: adapters/graphql-py/test_adapter.py
"""Integration tests for the ANIP GraphQL adapter.

Usage:
    python test_adapter.py http://localhost:9100
"""

from __future__ import annotations

import asyncio
import sys

import httpx

# ---------------------------------------------------------------------------
# Test helpers
# ---------------------------------------------------------------------------

_pass_count = 0
_fail_count = 0


def _ok(label: str) -> None:
    global _pass_count
    _pass_count += 1
    print(f"  PASS  {label}")


def _fail(label: str, detail: str = "") -> None:
    global _fail_count
    _fail_count += 1
    msg = f"  FAIL  {label}"
    if detail:
        msg += f"  — {detail}"
    print(msg)


def _assert(condition: bool, label: str, detail: str = "") -> None:
    if condition:
        _ok(label)
    else:
        _fail(label, detail)


# ---------------------------------------------------------------------------
# 1. Config defaults
# ---------------------------------------------------------------------------

def test_config_defaults() -> None:
    print("\n--- Config defaults ---")
    from anip_graphql_adapter.config import AdapterConfig

    cfg = AdapterConfig()
    _assert(cfg.anip_service_url == "http://localhost:8000", "default service URL")
    _assert(cfg.port == 3002, "default port")
    _assert(cfg.graphql.path == "/graphql", "default graphql path")
    _assert(cfg.graphql.playground is True, "default playground enabled")
    _assert(cfg.graphql.introspection is True, "default introspection enabled")


# ---------------------------------------------------------------------------
# 2. Discovery
# ---------------------------------------------------------------------------

async def test_discovery(anip_url: str) -> None:
    print("\n--- Discovery ---")
    from anip_graphql_adapter.discovery import discover_service

    service = await discover_service(anip_url)
    _assert(service.base_url != "", "base_url populated")
    _assert(service.protocol != "", "protocol populated")
    _assert(len(service.capabilities) > 0, "capabilities discovered")
    _assert("search_flights" in service.capabilities, "search_flights found")
    _assert("book_flight" in service.capabilities, "book_flight found")

    sf = service.capabilities["search_flights"]
    _assert(sf.side_effect == "read", "search_flights is read")
    _assert(sf.financial is False, "search_flights not financial")

    bf = service.capabilities["book_flight"]
    _assert(bf.side_effect in ("write", "irreversible"), "book_flight has side effect")
    _assert(bf.financial is True, "book_flight is financial")
    _assert(len(bf.requires) > 0, "book_flight has prerequisites")


# ---------------------------------------------------------------------------
# 3. Schema generation
# ---------------------------------------------------------------------------

async def test_schema_generation(anip_url: str) -> None:
    print("\n--- Schema generation ---")
    from anip_graphql_adapter.discovery import discover_service
    from anip_graphql_adapter.translation import generate_schema

    service = await discover_service(anip_url)
    sdl = generate_schema(service)

    # Directives present
    _assert("directive @anipSideEffect" in sdl, "SDL has @anipSideEffect directive")
    _assert("directive @anipCost" in sdl, "SDL has @anipCost directive")
    _assert("directive @anipRequires" in sdl, "SDL has @anipRequires directive")
    _assert("directive @anipScope" in sdl, "SDL has @anipScope directive")

    # Shared types
    _assert("scalar JSON" in sdl, "SDL has JSON scalar")
    _assert("type CostActual" in sdl, "SDL has CostActual type")
    _assert("type ANIPFailure" in sdl, "SDL has ANIPFailure type")
    _assert("type Resolution" in sdl, "SDL has Resolution type")

    # Query and Mutation
    _assert("type Query" in sdl, "SDL has Query type")
    _assert("type Mutation" in sdl, "SDL has Mutation type")

    # camelCase names
    _assert("searchFlights" in sdl, "SDL uses camelCase searchFlights")
    _assert("bookFlight" in sdl, "SDL uses camelCase bookFlight")

    # Result types (PascalCase)
    _assert("SearchFlightsResult" in sdl, "SDL has SearchFlightsResult type")
    _assert("BookFlightResult" in sdl, "SDL has BookFlightResult type")

    # searchFlights is in Query (read), bookFlight is in Mutation
    # Find the Query block and check searchFlights is there
    query_block = sdl[sdl.index("type Query"):sdl.index("}", sdl.index("type Query")) + 1]
    _assert("searchFlights" in query_block, "searchFlights is in Query type")

    mutation_block = sdl[sdl.index("type Mutation"):sdl.index("}", sdl.index("type Mutation")) + 1]
    _assert("bookFlight" in mutation_block, "bookFlight is in Mutation type")

    # Directives applied to fields
    _assert("@anipSideEffect" in sdl.split("type Query")[1], "fields have @anipSideEffect")


# ---------------------------------------------------------------------------
# 4. Full server (httpx ASGITransport)
# ---------------------------------------------------------------------------

async def test_server(anip_url: str) -> None:
    print("\n--- Server (ASGI) ---")
    from anip_graphql_adapter.config import AdapterConfig
    from anip_graphql_adapter.server import build_app

    config = AdapterConfig(anip_service_url=anip_url)
    app = await build_app(config)

    # Pre-issue a signed token outside the ASGI transport block
    from anip_graphql_adapter.discovery import discover_service

    service = await discover_service(anip_url)
    async with httpx.AsyncClient(timeout=30) as client:
        token_resp = await client.post(
            service.endpoints["tokens"],
            json={
                "subject": "adapter:anip-graphql-adapter",
                "scope": ["*"],
                "capability": "search_flights",
            },
            headers={"Authorization": "Bearer demo-human-key"},
        )
        token_resp.raise_for_status()
        token_data = token_resp.json()
        signed_token = token_data["token"]

    transport = httpx.ASGITransport(app=app)
    async with httpx.AsyncClient(transport=transport, base_url="http://test") as client:
        # GET /schema.graphql returns SDL
        resp = await client.get("/schema.graphql")
        _assert(resp.status_code == 200, "GET /schema.graphql returns 200")
        sdl = resp.text
        _assert("type Query" in sdl, "schema.graphql contains Query type")
        _assert("directive @anipSideEffect" in sdl, "schema.graphql has directives")

        # Missing credentials test: POST /graphql without auth headers
        resp = await client.post(
            "/graphql/",
            json={
                "query": """
                    query {
                        searchFlights(
                            origin: "SEA"
                            destination: "SFO"
                            date: "2026-03-10"
                        ) {
                            success
                            result
                            failure {
                                type
                                detail
                            }
                        }
                    }
                """,
            },
        )
        _assert(resp.status_code == 200, "POST /graphql without creds returns 200 (GraphQL layer)")
        data = resp.json()
        search_data = data.get("data", {}).get("searchFlights", {})
        _assert(search_data.get("success") is False, "no-creds searchFlights success is false")
        _assert(
            search_data.get("failure", {}).get("type") == "missing_credentials",
            f"no-creds failure type is missing_credentials: {search_data.get('failure')}",
        )

        # API-key path: POST /graphql with X-ANIP-API-Key header
        resp = await client.post(
            "/graphql/",
            json={
                "query": """
                    query {
                        searchFlights(
                            origin: "SEA"
                            destination: "SFO"
                            date: "2026-03-10"
                        ) {
                            success
                            result
                        }
                    }
                """,
            },
            headers={"X-ANIP-API-Key": "demo-human-key"},
        )
        _assert(resp.status_code == 200, "POST /graphql searchFlights with API key returns 200")
        data = resp.json()
        _assert("errors" not in data or data["errors"] is None,
                f"searchFlights has no errors (api-key): {data.get('errors')}")
        search_data = data.get("data", {}).get("searchFlights", {})
        _assert(search_data.get("success") is True, "searchFlights success is true (api-key)")
        _assert(search_data.get("result") is not None, "searchFlights has result (api-key)")

        # Token path: POST /graphql with X-ANIP-Token header
        resp = await client.post(
            "/graphql/",
            json={
                "query": """
                    query {
                        searchFlights(
                            origin: "SEA"
                            destination: "SFO"
                            date: "2026-03-10"
                        ) {
                            success
                            result
                        }
                    }
                """,
            },
            headers={"X-ANIP-Token": signed_token},
        )
        _assert(resp.status_code == 200, "POST /graphql searchFlights with token returns 200")
        data = resp.json()
        _assert("errors" not in data or data["errors"] is None,
                f"searchFlights has no errors (token): {data.get('errors')}")
        search_data = data.get("data", {}).get("searchFlights", {})
        _assert(search_data.get("success") is True, "searchFlights success is true (token)")

        # POST /graphql with mutation { bookFlight(...) { success result costActual { ... } } }
        resp = await client.post(
            "/graphql/",
            json={
                "query": """
                    mutation {
                        bookFlight(
                            flightNumber: "AA100"
                            date: "2026-03-10"
                            passengers: 1
                        ) {
                            success
                            result
                            costActual {
                                financial {
                                    amount
                                    currency
                                }
                                varianceFromEstimate
                            }
                        }
                    }
                """,
            },
            headers={"X-ANIP-API-Key": "demo-human-key"},
        )
        _assert(resp.status_code == 200, "POST /graphql bookFlight returns 200")
        data = resp.json()
        _assert("errors" not in data or data["errors"] is None,
                f"bookFlight has no errors: {data.get('errors')}")
        book_data = data.get("data", {}).get("bookFlight", {})
        _assert("success" in book_data, "bookFlight has success key")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python test_adapter.py <anip-service-url>")
        print("Example: python test_adapter.py http://localhost:9100")
        sys.exit(1)

    anip_url = sys.argv[1]
    print(f"Testing ANIP GraphQL adapter against {anip_url}")

    # 1. Config (sync)
    test_config_defaults()

    # 2-4. Async tests
    asyncio.run(_run_async_tests(anip_url))

    # Summary
    print(f"\n{'='*50}")
    print(f"Results: {_pass_count} passed, {_fail_count} failed")
    if _fail_count > 0:
        sys.exit(1)
    print("All tests passed!")


async def _run_async_tests(anip_url: str) -> None:
    await test_discovery(anip_url)
    await test_schema_generation(anip_url)
    await test_server(anip_url)


if __name__ == "__main__":
    main()
